src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/FootnoteRepository.cs

```diff
@@ -19,40 +18,134 @@ public FootnoteRepository(StatisticsDbContext context)
+                {
+                    FilterItemId = filterItemId
+                }).ToList(),
+                Indicators = indicatorIds.Select(indicatorId => new IndicatorFootnote
+                {
+                    IndicatorId = indicatorId
+                }).ToList(),
+                Subjects = subjectIds.Select(subjectId => new SubjectFootnote
+                {
+                    SubjectId = subjectId
+                }).ToList()
+            };
+
+            // TODO EES-2971 We could link Footnote with ReleaseSubject to simplify getting all foonotes for a subject.
+            // Currently this involves getting all footnotes with criteria matching the subject directly or indirectly.
+            // This would also prevent removing a subject from a release until all of its footnote links are removed.
+            // Currently a footnote could remain linked to a release and a subject (either directly or indirectly via
+            // its FilterItems/FilterGroups/Filters/Indicators) even though no ReleaseSubject exists.
+
+            await _context.Footnote.AddAsync(footnote);
+            await _context.SaveChangesAsync();
+
+            return footnote;
+        }
+
+        public async Task<List<Footnote>> GetFilteredFootnotes(
             Guid releaseId,
             Guid subjectId,
             IEnumerable<Guid> filterItemIds,
             IEnumerable<Guid> indicatorIds)
         {
-            var releaseIdParam = new SqlParameter("releaseId", releaseId);
-            var subjectIdParam = new SqlParameter("subjectId", subjectId);
-            var filterItemListParam = CreateIdListType("filterItemList", filterItemIds);
-            var indicatorListParam = CreateIdListType("indicatorList", indicatorIds);
+            // List the id's of all filter items, filter groups, filters, indicators
+            // Filter out any of these which don't belong to the same subject
+            var filterItemsQueryable = _context.FilterItem
+                .Where(filterItem => filterItemIds.Contains(filterItem.Id)
+                                     && filterItem.FilterGroup.Filter.SubjectId == subjectId);
+
+            var filterItems = await filterItemsQueryable
+                .Select(filterItem => filterItem.Id)
+                .ToListAsync();
+
+            var filterGroups = await filterItemsQueryable
+                .Select(filterItem => filterItem.FilterGroupId)
+                .Distinct()
+                .ToListAsync();
+
+            var filters = await filterItemsQueryable
+                .Select(filterItem => filterItem.FilterGroup.FilterId)
+                .Distinct()
+                .ToListAsync();
+
+            var indicators = await _context.Indicator
+                .Where(indicator => indicatorIds.Contains(indicator.Id)
+                                    && indicator.IndicatorGroup.SubjectId == subjectId)
+                .Select(indicator => indicator.Id)
+                .ToListAsync();
```

Comment:
I don't think we can return early because if both collections are empty we still want to return footnotes that are applied directly to the subject (i.e. those that have a `SubjectFootnote` matching the `subjectId` parameter).

I will add some unit tests to check it works in this case though, thanks.

Reply:
Update: we can return early by creating a branch in the code similar to how you say, checking `if (!filterIdsResult.Any() && !indicatorIdsSanitised.Any())`

It returns footnotes that applicable directly to a subject.

```
            // If the list of filter and indicator id's are empty,
            // return early with only footnotes that apply directly to this subject
            if (!filterIdsResult.Any() && !indicatorIdsSanitised.Any())
            {
                return await footnotesQueryable.Where(footnote =>
                        footnote.Subjects.Any(subjectFootnote => subjectFootnote.SubjectId == subjectId)

                        // We don't allow a footnote to apply to directly to a subject and to specific subject data
                        // of the same subject at the same time, so these additional conditions aren't strictly necessary
                        && footnote.FilterItems.All(itemFootnote =>
                            itemFootnote.FilterItem.FilterGroup.Filter.SubjectId != subjectId)
                        && footnote.FilterGroups.All(filterGroupFootnote =>
                            filterGroupFootnote.FilterGroup.Filter.SubjectId != subjectId)
                        && footnote.Filters.All(filterFootnote =>
                            filterFootnote.Filter.SubjectId != subjectId)
                        && footnote.Indicators.All(indicatorFootnote =>
                            indicatorFootnote.Indicator.IndicatorGroup.SubjectId != subjectId))
                    .OrderBy(footnote => footnote.Order)
                    .ThenBy(footnote => footnote.Id)
                    .ToListAsync();
            }
```